packages/sync/src/queue/index.ts

```diff
@@ -0,0 +1,64 @@
+import PQueue from 'p-queue'
+import { ConsecutiveList } from './consecutive-list'
+
+export { ConsecutiveList }
+
+export type SyncQueueOptions = {
+  setCursor?: (cursor: number) => Promise<void>
+  concurrency?: number
+  startCursor?: number
+}
+
+// A queue with arbitrarily many partitions, each processing work sequentially.
+// Partitions are created lazily and taken out of memory when they go idle.
+export class SyncQueue {
+  consecutive = new ConsecutiveList<number>()
+  mainQueue: PQueue
+  partitions = new Map<string, PQueue>()
+  cursor: number
+
+  constructor(public opts: SyncQueueOptions = {}) {
+    this.mainQueue = new PQueue({ concurrency: opts.concurrency ?? Infinity })
+    this.cursor = opts.startCursor ?? 0
+  }
+
+  async addTask(partitionId: string, task: () => Promise<void>) {
+    if (this.mainQueue.isPaused) return
```

Comment:
I wonder if we want to hoist this up into `trackEvt` so that the consecutive/cursor logic, etc. doesn't get triggered during destroy.

Reply:
Oh I see, you want `addTask` to be public API so it isn't necessarily guarded.  Maybe both places then?

Reply:
yeah good call